### ISSUE TITLE
Is rmlint still maintained?

### ISSUE THREAD

**Issue by StatusCode404** [sahib/rmlint#670]
Hi All,
Just checking (again but this year) if rmlint is still maintained and supported by anyone?  

There aren't that many responses to issues and there hasn't been a new tagged release since August last year.

**Comment by yarikoptic** [comment]
Ping on this @sahib and @cebtenzzre , a number of PRs send to be trivial and could be merged really fast and please all rmlint users. Got here after seeing @mih showing rmlint in action to dedicate annex keys. 

Cheers and let me know if we could be of some quick help ;-)

**Comment by sahib** [comment]
Original author here. I gave over maintenance to @cebtenzzre some time ago. As far as I can tell he's not responding either, which I can't blame him for. Open source work is seldom rewarding. I can give somebody else access to the repo to sift through the PRs, but this person should have a track of doing some open source work already, as I don't want another [xz incident](https://en.wikipedia.org/wiki/XZ_Utils_backdoor).

This person won't be me though since I don't posess the time and motivation to do so and also would write rmlint a lot differently today. Please answer in this issue if somebody steps up to do the job.

**Comment by yarikoptic** [comment]
I could probably give a hand with some trivial PRs and issues triage, may be an occasional upload to Debian, but not much beyond that. I have some track of FOSS development/maintenance (and also unfortunately abandoning as well ;) )

**Comment by sahib** [comment]
@yarikoptic Much appreciated. I would prefer if there's one additional person to reduce the risk of getting into the current situation.

**Comment by CodingWithAnxiety** [comment]
Hey there. 

Need another person? I'm an arch Linux user and would be available to help maintain.

**Comment by sahib** [comment]
> Hey there.
> 
> Need another person? I'm an arch Linux user and would be available to help maintain.

Cool! Thanks for raising your hand. Do you have some experience maintaining C applications? I see some Python experience which will be helpful for the test suite and UI.

@yarikoptic Still in? If yes I could give you guys access.

@cebtenzzre Please raise your hand if you do not want your access to be revoked.

**Comment by yarikoptic** [comment]
yes, but only to a very limited degree as described above.

**Comment by fermino** [comment]
Hey guys! Another arch linux user here (impressive what a deleted AUR package can do!). 
I'm willing to put in some time helping maintain this, maybe reviewing PRs and helping keep the test suite running :)

As a side note, I think that probably the sanest idea for now is to try keep things working and focus on fixing known bugs rather than trying to add new features, mostly because we're all new on this project :)

**Comment by yarikoptic** [comment]
also might be worth for @sahib to establish some "gatekeeping" e.g. that every PR must be approved by some other contributor first to be able to merge. (although likely they might be not "hard enforced" or I am a super user everywhere... dang... but example could be https://github.com/citeproc-py/citeproc-py)

**Comment by fermino** [comment]
It's a good idea! A 2-3 contributor approval + CI required to pass in the repo should a pretty robust filter.

**Comment by sahib** [comment]
I did setup those rules for `master` and `develop`. I would recommend though that there is only 1 required approval as, this can easily lead to stagnation otherwise. I also recommend to enable a positive CI check before merging, but this requires some work first sd the original CI site (TravisCI) vanished. 

@yarikoptic @CodingWithAnxiety @fermino: You should have collaboration invites now.

**Comment by fermino** [comment]
Awesome, thank you!

Regarding CI, I will look into that. I'm guessing it shouldn't be too hard to migrate it from Travis. Free github action minutes should probably be plenty for now!

**Comment by CodingWithAnxiety** [comment]
> > Hey there.
> > 
> > Need another person? I'm an arch Linux user and would be available to help maintain.
> 
> Cool! Thanks for raising your hand. Do you have some experience maintaining C applications? I see some Python experience which will be helpful for the test suite and UI.
> 
> @yarikoptic Still in? If yes I could give you guys access.
> 
> @cebtenzzre Please raise your hand if you do not want your access to be revoked.

Hi, 

I mostly have python experience under my belt, though I'm still learning C and C++. I'd mostly be interested in helping testing and squashing bugs. 

I'll keep my eyes on PRs and issues and see if I can't occasionally lend out a hand. 

I will accept the invention once I am home. <3

**Comment by sahib** [comment]
> @sahib probably a silly question but anyways: master is the latest branch, right? (Just making sure because I see the develop branch has other stuff but it's one year behind master).

develop is supposed to be the current working version with newest features and fixes. master is usually the one with the latest stable, released software. PRs would go to develop first, on release you rebase or merge to master. You are of course free to use a different branching model, but I think it is worth to revive and streamline the develop branch.

**Comment by fermino** [comment]
@sahib thanks for the info! I'm trying to figure out what to do with develop, mostly because I would not like to ship and release something not deemed stable (given user data is at stake :p). 

I see that most of the commits (or at least the ones I looked up) are new features, am I right? So in that case maybe the best way would be to start off master (specially about some build fixes for rolling release distros I've been looking at) and then go about integrating the things from develop to master. Any thoughts?

**Comment by sahib** [comment]
@fermino Sorry, bit late. Yes, seems like most features landed on develop, but some fixes are also on master, so the two need to be merged. First step would be to put this merged state on a separate branch, as most people compiling from source will master, but the [docs](https://rmlint.readthedocs.io/en/master/install.html#compilation) mentions develop. Once that new branch seems stable it can be moved to develop.

Upstream distros will not update until a new tag/release is pushed.

**Comment by sahib** [comment]
> > would write rmlint a lot differently today
> 
> This may be off-topic here, but I would be very interested to have a bit more detail on what rmlint could have been if you had started it in 2025.
> 
> And thanks a lot for rmlint, this is a useful software that I trust and have found useful. I also wrote my own software from scratch several times that share similar features. That is why I am interested in your own feedback after several years of experience of a larger and well-written project like rmlint.


Hmm, this probably would deserve a longer post, but here's what comes to mind:

- Do one thing well: Remove all the other lint stuff, like empty files, nonstripped etc. This appeared useful to me back then, but is more annoying in hindsight,  especially since it makes the implementation harder.
- Also allow integration for tools that find similar images, or other specialized use cases  (i.e make the hashing-core exchangeable in the architecture)
- Do not offer a UI: That was only because I got interested in GTK and cairo if I'm honest.  I do not mind it as separate project, but this attracts the wrong user base to a power tool.
- Only offer one checksum implementation and remove paranoia mode.  Modern checksums like blake are fast enough and very sturdy. There is really no need in   having an additional paranoia mode, considering it made things rather complex.
- Remove xattr support. That stuff never worked right.
- Heavily reduce the amount of knobs in the command line interface. More is not better.  Especially if you run out of letters.
- Base the new implementation based on io_uring. The current multi-threaded hasher has some needless complexity.
- Write it in a modern language that allows easy cross-compilation.  This would reduce bugs, make developer life less miserable during debugging and speed would be pretty much the same with Go/Rust/Zig. I would pick Go. We had so many bugs because we made the memory management awkward and I want that life time back.  I only knew C back then and Go was not yet available.
- Less outputs. The progressbar should be a proper ncurses or bubbletea one.  No python output or CSV. There are tools for that.
- Write the benchmarks before optimizing. ;-) This has too often turned to an exercise in optimization uncritical paths.  Also do not care so much about rotational disk anymore. They are already an edge case now, but have been the norm back then.

There were good ideas though:

- Keeping it non-interactive.
- Merging directories (should be default)
- Have a JSON output for scripting
- Replay as an idea is actually not bad.
- The test suite using black box tests of the actual binary is nice.
- Distinction between original and duplicate and tooling to decide.

Maybe this can also serve as inspiration for the current maintainers.

**Comment by misieck** [comment]
I like the --replay ability, but it should be the main modus rather than an odd add-on. Id like the app to always produce a database at first scan, that can be use for subsequent clean passes. Reasoning is the speed of operation.

Otherwise as of today rmlint is not maintained, and contains data-destroying bugs, see #672. Thats bad enough it should be removed from distro repositories.

**Comment by StatusCode404** [comment]
> I like the --replay ability, but it should be the main modus rather than an odd add-on. Id like the app to always produce a database at first scan, that can be use for subsequent clean passes. Reasoning is the speed of operation.
> 
> Otherwise as of today rmlint is not maintained, and contains data-destroying bugs, see [#672](https://github.com/sahib/rmlint/issues/672). Thats bad enough it should be removed from distro repositories.

I've never used "--replay".
But it is clearly stated in the rmlint manual page that it will not honour hardlinks if you choose --replay.
https://github.com/sahib/rmlint/issues/672#issuecomment-2638201349

```
By design, some options will not have any effect. Those are:

    --followlinks
    --algorithm
    --paranoid
    --clamp-low
    --hardlinked
    --write-unfinished
    … and all other caching options below.

```
Read the manual and you won't destroy your data @misieck  by eagerly using an option you don't fully understand.

**Comment by misieck** [comment]
No,  --hardlinked, means report hard links as duplicates. The negative of that is not to report them as duplicates and thus preserve both files. 
 *rmlint deletes both files*. There is no world in which this could be by design. People are losing data with this.

**Comment by rsbrux** [comment]
>     * Do one thing well: Remove all the other lint stuff, like empty files, nonstripped etc. This appeared useful to me back then, but is more annoying in hindsight,  especially since it makes the implementation harder.
> 
I'm just a lurker, but heartily endorse this idea. I recently got caught by [this](https://github.com/sahib/rmlint/issues/649).
I wanted to dedup a USB backup drive from a Windows system on my Ubuntu box, and got 0 duplicates and thousands of empty files/directories, apparently due to "bad GIDs".
If I have correctly understood the empty file/directory handling, running the shell script would have deleted almost everything.

**Comment by RayOei** [comment]
@Sahib What is the status on the maintenance pick-up? 
I noticed @fermino 👏 being busy but also being blocked by required reviews?
I could provide some assistance but I am aware that my Github pedigree is not showing that much as I have mostly worked on/for closed source projects. I started in C/C++, but nowadays all over the place depending the need at hand (embedded/back-end/mobile).

**Comment by fermino** [comment]
Hey guys, been a bit busy lately :)

@sahib I agree with all the ideas you mentioned up there. Would you be against simplifying this codebase? Or you think it would be better to start it from scratch in another language? (Personally it would be a great time to dive on zig but Go would be great too). I'm mentioning this because I don't think I understand understand the current codebase enough to replicate it without messing up on some of the same things 😆

I'm thinking about maybe thinking of it as a library and then a separate CLI interface which would allow for a GUI in the future. 

Also, having a good user base testing a new implementation of the project would be great, so if people are willing to chime in with the most used features and what pitfalls they see I'd be willing to start a rewrite on my free time. 

Anyways, just some random thoughs as I'm reading this :)

**Comment by sahib** [comment]
@fermino I would advise against rewriting from scratch. That would be something for a successor of `rmlint`, that has not the restrictions imposed by following how `rmlint does things. But for this projects the value comes from a user base relying on it - even on the weirder features. From my point of view, the most important thing right now would be a release that fixes some of those dangerous and outstanding bugs. This would greatly help the user base and reduce maintenance effort.

@RayOei Sure, the more people the better I guess. 😄  Please pick a ticket/review to work on and once you figured out a fix I will give you access.

@CodingWithAnxiety @yarikoptic Reminder that you still have collaborator access. Since you did not pick something to work on yet I would remove access after some days, unless you tell me not to.

Thanks to anyone helping.

**Comment by RayOei** [comment]
Hi @sahib, first action was to check/comment a PR (#678) as that one is very simple 😁

**Comment by RayOei** [comment]
Hi @sahib, FYI: I added #683 and I think I have a fix for #673 too (needs more checking though).

**Comment by sahib** [comment]
@RayOei I gave you access now. Please coordinate with the others a bit how you want to work (i.e. when to push to `develop`, plan what is next, when to close a PR, what would be needed for a release, ...). 👍

**Comment by vassilit** [comment]
> [@RayOei](https://github.com/RayOei) I gave you access now. Please coordinate with the others a bit how you want to work (i.e. when to push to `develop`, plan what is next, when to close a PR, what would be needed for a release, ...). 👍

I've pushed a PR yesterday. I wanted to follow what d915e53180587a25ad483485a087c672a436d57b says, but the develop branch has split at 0dbeefe7c73668142d2ec8e1ff3660d2403083ed.
So I pushed against master, as this is a data-loss avoiding hotfix.

I have merged master into develop and resolved conflicts manually on my local repo but I need to do some tests before pushing a PR for it. We would then need to merge develop into master.

**Comment by vassilit** [comment]
> > > would write rmlint a lot differently today
> Hmm, this probably would deserve a longer post, but here's what comes to mind:

Thank you for your detailed feedback.

I wrote a “similar” (but quite simpler) software in C for my own use, based on a different principle: I index the hashes of all files from different machines into a SQlite database (a bit like BSD mtree) and then I scan the database to find duplicates. The search for duplicate directories works fine. It also has the advantage for me of indexing files in archives. Unfortunately, it still has some quirks (optimizations I have to remove to cover corner cases) that make it difficult to use or even dangerous for someone unfamiliar with its code, that's why I haven't released it yet. I'll possibly do it this year.
The software that indexes and the one that destroys are two separate software and the paranoia mode (which compares bit by bit the files whose hashes have been detected as identical) exists more to protect me against possible bugs in the indexing software than against hash collisions.  I could use rmlint as well as a second opinion to be safe.

Also, I still use rmlint sometimes for local cleanups and I find it sad that it is disappearing from distributions. Also, I would like to continue to trust it with my data as much as I used to.
I got a bit familiar with the code yesterday while fixing a bug, and I find it nice.

That's why I also offer my help to maintain rmlint, which has helped me for many years.
The goal would be to merge the two conflicting branches, merge all bugfixes, clean up the tests in coordination with the other maintainers, check compatibility with the libraries currently offered in modern distributions and be ready to release a new (and final?) stable version. At least a base to then possibly add features.

**Comment by vassilit** [comment]
> > > would write rmlint a lot differently today
> Hmm, this probably would deserve a longer post, but here's what comes to mind:

@sahib, thank you for your detailed feedback.

I wrote a “similar” (but quite simpler) software in C for my own use, based on a different principle: I index the hashes of all files from different machines into a SQlite database (a bit like BSD mtree) and then I scan the database to find duplicates. The search for duplicate directories works fine. It also has the advantage for me of indexing files in archives. Unfortunately, it still has some quirks (optimizations I have to remove to cover corner cases) that make it difficult to use or even dangerous for someone unfamiliar with its code, that's why I haven't released it yet. I'll possibly do it this year.
The software that indexes and the one that destroys are two separate software and the paranoia mode (which compares bit by bit the files whose hashes have been detected as identical) exists more to protect me against possible bugs in the indexing software than against hash collisions.  I could use rmlint as well as a second opinion to be safe.

Also, I still use rmlint sometimes for local cleanups and I find it sad that it is disappearing from distributions. Also, I would like to continue to trust it with my data as much as I used to.
I got a bit familiar with the code yesterday while fixing a bug, and I find it nice.

That's why I also offer my help to maintain rmlint, which has helped me for many years.
The goal would be to merge the two conflicting branches, merge all bugfixes, clean up the tests in coordination with the other maintainers, check compatibility with the libraries currently offered in modern distributions and be ready to release a new (and final?) stable version. At least a base to then possibly add features.

**Comment by RayOei** [comment]
> I gave you access now.

Thanks!

**Comment by RayOei** [comment]
> I have merged master into develop and resolved conflicts manually

Hi @vassilit, I was playing around with that myself as well. At this point I believe we better get more acquainted with the code base before we try on a task to have `develop` be usable and reliable again. E.g. that you know that it is doing what it is supposed to do. A bit sad it was abandoned like that...

I think it may be a better choice to introduce a new `develop` and start working on feature/fix branches there and get back into a more sane git flow. E.g. not pushing to `master` all the time. But I guess that is something we should discuss with @fermino as well.
I would welcome your contribution too, the more the merrier, and there is enough to cleanup and improve. And great you looked at #672 as that was the next on my agenda to look at.

**Comment by RayOei** [comment]
Apart from fixing/updating the tests. That should be the priority at this point before we tinker too much, at least my 2 cnt 😁

**Comment by sahib** [comment]
Glad this is gaining some traction.

@vassilit I saw you working on issues quite a bit already, therefore you also have access now.

**Comment by vassilit** [comment]
> > I have merged master into develop and resolved conflicts manually
> Hi [@vassilit](https://github.com/vassilit), I was playing around with that myself as well. At this point I believe we better get more acquainted with the code base before we try on a task to have `develop` be usable and reliable again. E.g. that you know that it is doing what it is supposed to do.

IMHO, the changes are not so drastic between develop and master, and they don't touch core functionalities.
The code base is quite well-written here and with that amount of libs and features used, it could have been much worse. The authors did a great job. Even with using funny names like parrot and cage (and even `_tail`  fits in the theme, even if it's from Glibc ;) it's very readable even when there are no comments. The GNU styling is not up to my taste, but that's a detail.

> I think it may be a better choice to introduce a new `develop` and start working on feature/fix branches there and get back into a more sane git flow. 

I still think we should merge master into develop and then develop into master. After that, we should merge PR into develop.
However, an alternative would be to open PRs for every commits on master to be merged into develop and cherry-pick them one by one.

> And great you looked at [#672](https://github.com/sahib/rmlint/issues/672) as that was the next on my agenda to look at.

Yes, this is a data-loss issue that's why I merged it into master as an hotfix.

**Comment by RayOei** [comment]
> I still think we should merge master into develop and then develop into master. After that, we should merge PR into develop.

I agree 😁 And you have clearly already a better view of the code base and `develop` than I have at this point.

**Comment by RayOei** [comment]
@fermino @vassilit As to the main question in _this_ issue: should/can we project a new release and signal that things are getting on the road again? 
I am not sure `Projects` is a space for that (I don't have permissions to do anything with it)? 
But some coordination and communication would be useful?

**Comment by vassilit** [comment]
> [@fermino](https://github.com/fermino) [@vassilit](https://github.com/vassilit) As to the main question in _this_ issue: should/can we project a new release and signal that things are getting on the road again?

IMHO, a new release would be welcome. Nothing fancy, just bugfixes, already-merged features and compatibility with the modern toolchain distributions uses so that they could continue distribute the package and reduce the number of patch and therefore own maintenance they have to do on their side.

We also have to define a target regarding libc/kernel/GLib2 compatibility. Because fixing for modern toolchain and libs while keeping backward-compatibility (i.e. if someone wants to compile and use rmlint on a RHEL5) is sometimes difficult.

**Comment by RayOei** [comment]
Let's split those in separate issues and continue specific discussion there? 
And close this one?

**Comment by vassilit** [comment]
> [@fermino](https://github.com/fermino) [@vassilit](https://github.com/vassilit) As to the main question in _this_ issue: should/can we project a new release and signal that things are getting on the road again?

Probably even two releases. 2.10.3 after cleaning and 2.11.0 after merging new features.
@sahib when you have time, could you find a name for 2.10.3 and add "2.11.0 Evolving Echidna" to Milestones ? The latest version in Milestones is 2.8.0.

**Comment by RayOei** [comment]
There is a (old) pinned #510 issue on top of the issues page referring to v2.11.
Shall we repurpose/update that one and add a v2.10.3 todo issue as well? That way it will be a bit easier to track what is planned and what the status is.

**Comment by vassilit** [comment]
And/or using Milestones for that, which is IMHO clearer for me. But if you like a TODO list in an issue, that is good as well :)

**Comment by RayOei** [comment]
Now we have them 😁 

Milestone [2.10.3](https://github.com/sahib/rmlint/milestone/8)
